File: web/frontend/src/api/graph.ts
import { api } from './client'
import type { GraphGenerateRequest, GraphGenerateResponse, GraphUploadResponse } from './types'

export const generateGraph = (req: GraphGenerateRequest) =>
  api.post<GraphGenerateResponse>('/graph/generate', req)

export const getGraph = (graphId: string) =>
  api.get<{ graph_id: string; elements: unknown[] }>(`/graph/${graphId}`)

export const uploadGraphs = async (files: File[]): Promise<GraphUploadResponse> => {
  const formData = new FormData()
  for (const file of files) formData.append('files', file)
  const res = await fetch('/api/graph/upload', { method: 'POST', body: formData })
  if (!res.ok) {
    const detail = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(detail.detail ?? res.statusText)
  }
  return res.json()
}

File: web/backend/routers/graph.py
"""Graph generation and retrieval endpoints."""
import asyncio
from typing import List

from fastapi import APIRouter, File, HTTPException, UploadFile

from web.backend.models.graph_models import (
    GraphGenerateRequest,
    GraphGenerateResponse,
    GraphUploadResponse,
)
from web.backend.services import graph_service, graph_store, serialization, upload_service

router = APIRouter()


@router.post("/generate", response_model=GraphGenerateResponse)
async def generate_graph(request: GraphGenerateRequest) -> GraphGenerateResponse:
    try:
        graph_id, elements, stats = await asyncio.to_thread(graph_service.generate, request)
        return GraphGenerateResponse(graph_id=graph_id, elements=elements, stats=stats)
    except Exception as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.post("/upload", response_model=GraphUploadResponse)
async def upload_graph(files: List[UploadFile] = File(...)) -> GraphUploadResponse:
    try:
        graph_id, elements, stats, file_count, folder_path, warnings = (
            await asyncio.to_thread(upload_service.upload, files)
        )
        return GraphUploadResponse(
            graph_id=graph_id,
            elements=elements,
            stats=stats,
            file_count=file_count,
            folder_path=folder_path,
            warnings=warnings,
        )
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.get("/{graph_id}")
async def get_graph(graph_id: str) -> dict:
    try:
        graph = graph_store.get(graph_id)
    except KeyError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    elements = serialization.graph_to_elements(graph)
    return {"graph_id": graph_id, "elements": [e.model_dump() for e in elements]}
